project/Boilerplate.scala

```diff
@@ -0,0 +1,192 @@
+   */
+  def gen(dir : File) = for (t <- templates) yield {
+    val tgtFile = t.filename(dir)
+    IO.write(tgtFile, t.body)
+    tgtFile
+  }
+
+  /**
+   * Return a sequence of the generated test files. As a side-effect, it actually generates them...
+   */
+  def genTests(dir: File): Seq[File] = testTemplates.map { template =>
+    val tgtFile = template.filename(dir)
+    IO.write(tgtFile, template.body)
+    tgtFile
+  }
+
+  class TemplateVals(val arity: Int) {
+    val synTypes = (0 until arity).map(n => s"A$n")
+    val synVals = (0 until arity).map(n => s"a$n")
+
+    val `A..N` = synTypes.mkString(", ")
+  }
+
+  /**
+   * Blocks in the templates below use a custom interpolator, combined with post-processing to
+   * produce the body.
+   *
+   * - The contents of the `header` val is output first
+   * - Then the first block of lines beginning with '|'
+   * - Then the block of lines beginning with '-' is replicated once for each arity,
+   *   with the `templateVals` already pre-populated with relevant relevant vals for that arity
+   * - Then the last block of lines prefixed with '|'
+   *
+   * The block otherwise behaves as a standard interpolated string with regards to variable
+   * substitution.
+   */
+  trait Template {
+    def filename(root: File): File
+    def content(tv: TemplateVals): String
+    def range: IndexedSeq[Int] = 1 to maxArity
+    def body: String = {
+      val headerLines = header.split('\n')
+      val raw = range.map(n => content(new TemplateVals(n)).split('\n').filterNot(_.isEmpty))
+      val preBody = raw.head.takeWhile(_.startsWith("|")).map(_.tail)
+      val instances = raw.flatMap(_.filter(_.startsWith("-")).map(_.tail))
+      val postBody = raw.head.dropWhile(_.startsWith("|")).dropWhile(_.startsWith("-")).map(_.tail)
+      (headerLines ++ preBody ++ instances ++ postBody).mkString("\n")
+    }
+  }
+
+  object GenProductReaders extends Template {
+    def filename(root: sbt.File) = root / "pureconfig" / "ProductReaders.scala"
+
+    def content(tv: TemplateVals): String = {
+      import tv._
+
+      val paramNames = synTypes.map(tpe => s"key$tpe: String").mkString(", ")
+      val readerInstances = synTypes.map(tpe => s"reader$tpe: ConfigReader[$tpe]").mkString(", ")
+
+      val results = synTypes.zip(synVals).map { case (tpe, v) =>
+        s"-        val ${v}Result = if (reader$tpe.isInstanceOf[AllowMissingKey]) reader$tpe.from(objCur.atKeyOrUndefined(key$tpe)) else objCur.atKey(key$tpe).right.flatMap(reader$tpe.from)"
+      }.mkString("\n")
+
+      val curriedTypes = synTypes.mkString(" => ")
+
+      val resultCombination = synVals.foldLeft("baseF") { case (curr, v) =>
+        s"ConvertHelpers.combineResults($curr, ${v}Result)(_(_))"
+      }
+
+      block"""
+        |package pureconfig
+        |
+        |import pureconfig.error._
+        |
+        |private[pureconfig] trait ProductReaders {
```

Comment:
Why the need to be private? Same thing for writers.

Reply:
My idea was having these traits providing methods as if they were defined in the `ConfigReader` and `ConfigWriter` objects. That's why I made them private and didn't provide a public object extending the trait. There's no functional need for this to be private other than hiding it from the outside.

Reply:
I didn't have a strong feeling towards having this private, so I ended up dropping it.